core/string/ustring.cpp

```diff
@@ -1644,6 +1645,32 @@ String String::hex_encode_buffer(const uint8_t *p_buffer, int p_len) {
 	return ret;
 }
 
+Vector<uint8_t> String::hex_decode() const {
+	const String *s = this;
+	Vector<uint8_t> out;
+	ERR_FAIL_COND_V_MSG(s->size() % 2 == 0, out, "Hexadecimal string of uneven length.");
+	int len = s->size() / 2;
+	out.resize(len);
+	for (int i = 0; i < len; i++) {
+		uint8_t hex = 0;
+		for (int j = i * 2; j < (i * 2) + 2; j++) {
+			char c = lower_case(s->get(j));
+			uint8_t n;
+			if (is_digit(c)) {
+				n = c - '0';
+			} else if (c >= 'a' && c <= 'f') {
+				n = (c - 'a') + 10;
+			} else {
+				ERR_FAIL_COND_V_MSG(true, out, "Invalid hexadecimal notation character \"" + chr(c) + "\" in string \"" + *this + "\".");
```

Comment:
```suggestion
				ERR_FAIL_V_MSG(Vector<uint8_t>(), "Invalid hexadecimal notation character \"" + chr(s->get(j)) + "\".");
```

1. There is `ERR_FAIL_V_MSG` macro.
2. `out` may be changed, return `Vector<uint8_t>()` instead.
3. `c` is lowercased character, you need the original one.
4. The string may be too long, I think it's better not to зкште it.

Reply:
Your making the loop into 2 blocks of identical code.

Reply:
> Your making the loop into 2 blocks of identical code, and your making it more complex for no reason

I'll just quote the answer to me in a similar situation:

> Please be aware that shorter code is not always better code.

You have an unnecessary nested loop.

> 3) the original doesnt work.

https://github.com/godotengine/godot/blob/5f2137a05040beb84d14ca6dd465688d3917a3ee/core/string/ustring.cpp#L2160

https://github.com/godotengine/godot/blob/5f2137a05040beb84d14ca6dd465688d3917a3ee/core/string/ustring.cpp#L2167

If you `"FG".hex_decode()` you will get an error about `g`, not about `G`.

> 4) What is зкште?

Wrong keyboard layout.

Reply:
Yes, but this shorter code is the same speed as a nested loop

Reply:
The nested loop here is less readable than code duplication. Also, the nested loop adds an extra check for each character (`j < (i * 2) + 2`). Don't rely on a smart compiler if you can avoid extra steps and extra variables in a simple way.

I took a look at PHP's `hex2bin()` implementation for comparison. It also does not use a nested loop and there is code duplication. (In addition, bitwise shifts are used instead of multiplication and division.)

https://github.com/php/php-src/blob/368febbf89cd1124b4041b126152255bd3fc89f2/ext/standard/string.c#L81-L115

Reply:
okay, ill duplicate, but i cant use bitwise shifts and such cause then i cant handle errors.